cyral/resource_cyral_sidecar_listener.go

```diff
@@ -0,0 +1,363 @@
+				Type:        schema.TypeList,
+				Required:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			NetworkAddressKey: {
+				Description:  "The network address that the sidecar listens on.",
+				Type:         schema.TypeSet,
+				Optional:     true, // This field is required, due to ExactlyOneOf directive
+				ExactlyOneOf: []string{NetworkAddressKey},
+				MaxItems:     1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						HostKey: {
+							Description: "Host where the sidecar will listen for the given repository. Omit to listen on all interfaces.",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+						PortKey: {
+							Description: "Port where the sidecar will listen for the given repository.",
+							Type:        schema.TypeInt,
+							Required:    true,
+						},
+					},
+				},
+			},
+			MySQLSettingsKey: {
+				Description: "MySQL settings represents the listener settings for a [`mysql`, `galera`, `mariadb`] data repository.",
+				Type:        schema.TypeSet,
+				Optional:    true,
+				// Notice the MaxItems: 1 here. This ensures that the user can only specify one this block.
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						DbVersionKey: {
+							Description: "MySQL DB version. Required (and only relevant) for listeners of type `mysql`.",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+						CharacterSetKey: {
+							Description: "MySQL character set. Optional and only relevant for listeners of type `mysql`.",
+							Type:        schema.TypeString,
+							Optional:    true,
+							Computed:    true,
+						},
+					},
+				},
+			},
+			S3SettingsKey: {
+				Description: "S3 settings.",
+				Type:        schema.TypeSet,
+				Optional:    true,
+				// Notice the MaxItems: 1 here. This ensures that the user can only specify one this block.
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						ProxyModeKey: {
+							Description: "S3 proxy mode, only relevant for S3 listeners. Defaults to false.",
+							Type:        schema.TypeBool,
+							Optional:    true,
+						},
+					},
+				},
+			},
+			DynamoDbSettingsKey: {
+				Description: "DynamoDB settings.",
+				Type:        schema.TypeSet,
+				Optional:    true,
+				// Notice the MaxItems: 1 here. This ensures that the user can only specify one this block.
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						ProxyModeKey: {
+							Description: "DynamoDB proxy mode. Only relevant for listeners of type `dynamodb`. Note " +
+								"that `proxy_mode` must be set to `true` for listeners of type `dynamodb`. Defaults to false.",
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
```

Comment:
Not sure if I actually followed this description. If the ProxyMode must be set to true, then why do we need this argument? Seems to me that would be better to completely remove the DynamoDbSettings as theres no much to be configured in this case. @gengdahlCyral why is this field added to the Listener API? I might be missing something here

Reply:
Hi @VictorGFM 

> If the ProxyMode must be set to true, then why do we need this argument? 

Because it dictates how the sidecar will handle that specific TCP connection.


> Seems to me that would be better to completely remove the DynamoDbSettings as there's no much to be configured in this case.

I initially disagree, because even though this field must always be `true` today for DynamoDB, we may allow it to be false in the future and it means if we remove it now and re-add it later, that would bring an unnecessary change in the listener payload, especially because when this field is missing we would not now if we should consider the value to be false or true.


> @gengdahlCyral could you help give the context of why this field was added to the Listener API? I might be missing something here

It was added because the sidecar **supports** both proxy and non-proxy mode for DynamoDB and DynamoDB Streams. However, we are not allowing the configuration with `proxy == false` to be used in an initial moment.

I don't know if this is possible, but if we could have the default as `true` for DynamoDB that would be better, but I don't know if that is feasible here or if that is a good choice in the terraform context.

Reply:
@juniocezar I see, thanks for the context. Given that we might allow `ProxyMode` for dynamoDB to be set to `false` in the future we might want to keep this field then. Regarding the default value, I personally think that the resource should match the API behavior, so in this case - since the API requires the `ProxyMode` to be `true` - I would suggest making the dynamoDB `ProxyMode` a required field (`Required: true`) and keep the default value to `false`, as it is today in the API.

Reply:
@VictorGFM I am hesitant to make ProxyMode to required since it is not *strictly* required by the API defintion (see [here](https://github.com/cyralinc/wrapper-client-go/blob/137c6aed3247ca86e4d2988f5ac181d26f23c6af/messages/listeners.proto#L672)-- the validation is enforced in the go code, which is subject to change. However, I think I will go ahead and do it, since it is technically required by the API.

Reply:
@hbaackmann yea, I understand your point. From my perspective, keeping the ProxyMode optional for dynamoDB would be inconsistent with the current API behavior, given that today the Listener API requires the ProxyMode to be true for dynamoDB - even though its not explicitly defined in the proto design - and I'm usually in favor that the resource replicates the API behavior (If in the future we make this field optional in the API we should update the resource to match that). That said, I just wanted to share my thought about this here but I think its ok if we decide to keep the ProxyMode optional (even though I think its a bit inconsistent).